kinzhal-processor/src/jvmMain/kotlin/com/daugeldauge/kinzhal/processor/generation/AssistedFactoryGenerator.kt

```diff
@@ -0,0 +1,165 @@
+package com.daugeldauge.kinzhal.processor.generation
+
+import com.daugeldauge.kinzhal.annotations.Assisted
+import com.daugeldauge.kinzhal.processor.NonRecoverableProcessorException
+import com.daugeldauge.kinzhal.processor.model.AssistedFactoryType
+import com.daugeldauge.kinzhal.processor.model.FactoryBinding
+import com.daugeldauge.kinzhal.processor.model.Key
+import com.daugeldauge.kinzhal.processor.requireQualifiedName
+import com.daugeldauge.kinzhal.processor.toKey
+import com.google.devtools.ksp.processing.CodeGenerator
+import com.google.devtools.ksp.symbol.KSFunctionDeclaration
+import com.google.devtools.ksp.symbol.KSValueParameter
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.LambdaTypeName
+import com.squareup.kotlinpoet.ParameterSpec
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.ksp.toTypeName
+import com.squareup.kotlinpoet.withIndent
+
+internal fun generateAssistedFactory(
+    codeGenerator: CodeGenerator,
+    injectableKey: Key,
+    sourceDeclaration: KSFunctionDeclaration,
+    addCreateInstanceCall: CodeBlock.Builder.() -> Unit,
+    packageName: String,
+    factoryBaseName: String,
+    assistedFactoryType: AssistedFactoryType,
+): FactoryBinding {
+    class Dependency(
+        val name: String,
+        val key: Key,
+        val isAssisted: Boolean,
+        val parameter: KSValueParameter,
+    )
+
+    val dependencies = sourceDeclaration.parameters.map {
+        val isAssisted = it.isAssisted
+        Dependency(
+            name = if (isAssisted) {
+                it.name!!.asString()
+            } else {
+                "${it.name!!.asString()}Provider"
+            },
+            key = it.type.toKey(it.annotations),
+            isAssisted = isAssisted,
+            parameter = it
+        )
+    }
+
+    val assistedSourceParameters = dependencies
+        .asSequence()
+        .filter(Dependency::isAssisted)
+        .map { it.name to it.key.type }
+        .toList()
+
+    val assistedFactoryParameters = assistedFactoryType.factoryMethod.parameters
+        .map { it.name?.asString() to it.type.resolve() }
+
+    if (assistedSourceParameters != assistedFactoryParameters) {
+        throw NonRecoverableProcessorException(
+            "Factory method in @AssistedFactory must have assisted parameters by same names, types and in the same order as @AssistedInject constructor",
```

Comment:
I think I'd be a little better if this logic (and annotation matching above) was somewhere outside the generator.

Reply:
Done - extracted to a separate utility object `AssistedFactoryDependenciesResolver`